kahypar/partition/bin_packing/bin_packing.h

```diff
@@ -0,0 +1,458 @@
+        return _alg.binWeight(bin);
+      }
+
+      // ... and at the second level, the resulting k bins are packed into the final parts.
+      // Returns the partition mapping for the bins and a vector of the resulting part weights
+      std::pair<PartitionMapping, std::vector<HypernodeWeight>> applySecondLevel(const std::vector<HypernodeWeight>& max_allowed_part_weights,
+                                                                                 const std::vector<PartitionID>& num_bins_per_part) const {
+        ASSERT(num_bins_per_part.size() == max_allowed_part_weights.size());
+
+        const PartitionID num_parts = static_cast<PartitionID>(max_allowed_part_weights.size());
+        std::vector<PartitionID> bin_counts(max_allowed_part_weights.size(), 0);
+        PartitionMapping mapping = _bins_to_parts;
+
+        const HypernodeWeight max_part_weight = *std::max_element(max_allowed_part_weights.cbegin(),
+                                                max_allowed_part_weights.cend());
+        BPAlgorithm partition_packer(num_parts, max_part_weight);
+        for (PartitionID i = 0; i < num_parts; ++i) {
+          partition_packer.addWeight(i, max_part_weight - max_allowed_part_weights[i]);
+        }
+
+        for (PartitionID bin = 0; bin < _alg.numBins(); ++bin) {
+          if (mapping.isFixedBin(bin)) {
+            partition_packer.addWeight(mapping.binPartition(bin), _alg.binWeight(bin));
+          }
+        }
+
+        std::vector<PartitionID> kbins_descending(_alg.numBins());
+        std::iota(kbins_descending.begin(), kbins_descending.end(), 0);
+        std::sort(kbins_descending.begin(), kbins_descending.end(), [&](PartitionID i, PartitionID j) {
+          return _alg.binWeight(i) > _alg.binWeight(j);
+        });
+
+        for (const PartitionID& bin : kbins_descending) {
+          if (!mapping.isFixedBin(bin)) {
+            const PartitionID part = partition_packer.insertElement(_alg.binWeight(bin));
+            ASSERT((part >= 0) && (part < num_parts));
+
+            const size_t part_idx = static_cast<size_t>(part);
+            bin_counts[part_idx]++;
+            if (bin_counts[part_idx] >= num_bins_per_part[part_idx]) {
+              partition_packer.lockBin(part);
+            }
+
+            mapping.setPart(bin, part);
+          }
+        }
+
+        // collect part weights
+        std::vector<HypernodeWeight> part_weights;
+        part_weights.reserve(num_parts);
+        for (PartitionID i = 0; i < num_parts; ++i) {
+          part_weights.push_back(partition_packer.binWeight(i) + max_allowed_part_weights[i] - max_part_weight);
+        }
+
+        return {std::move(mapping), std::move(part_weights)};
+      }
+
+    private:
+      BPAlgorithm _alg;
+      PartitionMapping _bins_to_parts;
+  };
+
+  // Returns the current hypernodes sorted in descending order of weight.
+  static inline std::vector<HypernodeID> extractNodesWithDescendingWeight(const Hypergraph& hg) {
```

Comment:
consider adding a parameter that indicates wheater nodes with same weight should be random shuffled or not